File: packages/types/src/sealedbox.ts
import nacl from "tweetnacl";
import { blake2b } from "blakejs";

// TypeScript twin of cre/workflow/sealedbox.go -- must stay byte-for-byte compatible, since
// furnishers seal records here (browser or Node) and only the CRE enclave's Go implementation
// ever opens them. Verified interoperable in packages/types/src/sealedbox.test.ts and again
// cross-language in cre/workflow/sealedbox_test.go.

export interface KeyPair {
  publicKey: Uint8Array;
  secretKey: Uint8Array;
}

export function generateKeyPair(): KeyPair {
  const kp = nacl.box.keyPair();
  return { publicKey: kp.publicKey, secretKey: kp.secretKey };
}

function sealedBoxNonce(ephemeralPub: Uint8Array, recipientPub: Uint8Array): Uint8Array {
  const combined = new Uint8Array(ephemeralPub.length + recipientPub.length);
  combined.set(ephemeralPub, 0);
  combined.set(recipientPub, ephemeralPub.length);
  return blake2b(combined, undefined, 24);
}

/** libsodium crypto_box_seal: ephemeral keypair per message, sender needs only the recipient's public key. */
export function sealAnonymous(message: Uint8Array, recipientPublicKey: Uint8Array): Uint8Array {
  const ephemeral = nacl.box.keyPair();
  const nonce = sealedBoxNonce(ephemeral.publicKey, recipientPublicKey);
  const ciphertext = nacl.box(message, nonce, recipientPublicKey, ephemeral.secretKey);

  const sealed = new Uint8Array(ephemeral.publicKey.length + ciphertext.length);
  sealed.set(ephemeral.publicKey, 0);
  sealed.set(ciphertext, ephemeral.publicKey.length);
  return sealed;
}

/** Opens a crypto_box_seal ciphertext. Only ever needed here for tests -- production opening happens in the Go enclave. */
export function openAnonymous(sealed: Uint8Array, recipientPublicKey: Uint8Array, recipientSecretKey: Uint8Array): Uint8Array {
  const overhead = nacl.box.overheadLength;
  if (sealed.length < recipientPublicKey.length + overhead) {
    throw new Error("sealed box too short");
  }

  const ephemeralPub = sealed.slice(0, 32);
  const ciphertext = sealed.slice(32);
  const nonce = sealedBoxNonce(ephemeralPub, recipientPublicKey);

  const opened = nacl.box.open(ciphertext, nonce, ephemeralPub, recipientSecretKey);
  if (!opened) throw new Error("failed to open sealed box: authentication failed");
  return opened;
}

export function toHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

export function fromHex(hex: string): Uint8Array {
  const clean = hex.startsWith("0x") ? hex.slice(2) : hex;
  const bytes = new Uint8Array(clean.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(clean.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

File: cre/workflow/sealedbox.go
package workflow

import (
	"crypto/rand"
	"errors"

	"golang.org/x/crypto/blake2b"
	"golang.org/x/crypto/nacl/box"
)

// SealAnonymous implements libsodium's crypto_box_seal construction: an ephemeral X25519
// keypair is generated per message so the sender needs only the recipient's public key, and
// the recipient cannot identify the sender. This is what a furnisher uses to encrypt a
// position record to the CRE workflow's published public key (docs/architecture.md #3.3).
func SealAnonymous(message []byte, recipientPublicKey *[32]byte) ([]byte, error) {
	ephemeralPub, ephemeralPriv, err := box.GenerateKey(rand.Reader)
	if err != nil {
		return nil, err
	}

	nonce, err := sealedBoxNonce(ephemeralPub, recipientPublicKey)
	if err != nil {
		return nil, err
	}

	sealed := make([]byte, 0, 32+len(message)+box.Overhead)
	sealed = append(sealed, ephemeralPub[:]...)
	sealed = box.Seal(sealed, message, nonce, recipientPublicKey, ephemeralPriv)
	return sealed, nil
}

// OpenAnonymous decrypts a crypto_box_seal ciphertext. This is the only place the CRE
// workflow's private key is ever used -- it runs inside the enclave (docs/plan.md T-011).
func OpenAnonymous(sealed []byte, recipientPublicKey, recipientPrivateKey *[32]byte) ([]byte, error) {
	if len(sealed) < 32+box.Overhead {
		return nil, errors.New("sealed box too short")
	}

	var ephemeralPub [32]byte
	copy(ephemeralPub[:], sealed[:32])

	nonce, err := sealedBoxNonce(&ephemeralPub, recipientPublicKey)
	if err != nil {
		return nil, err
	}

	opened, ok := box.Open(nil, sealed[32:], nonce, &ephemeralPub, recipientPrivateKey)
	if !ok {
		return nil, errors.New("failed to open sealed box: authentication failed")
	}
	return opened, nil
}

// sealedBoxNonce matches libsodium: nonce = generichash(ephemeralPub || recipientPub, 24).
func sealedBoxNonce(ephemeralPub, recipientPub *[32]byte) (*[24]byte, error) {
	h, err := blake2b.New(24, nil)
	if err != nil {
		return nil, err
	}
	h.Write(ephemeralPub[:])
	h.Write(recipientPub[:])

	var nonce [24]byte
	copy(nonce[:], h.Sum(nil))
	return &nonce, nil
}
